sydent/config/server.py

```diff
@@ -0,0 +1,252 @@
+            self.general,
+            self.email,
+            self.database,
+            self.http,
+            self.sms,
+            self.crypto,
+        ]
+
+    def parse_config_file(self, config_file: str):
+        """Parse the given config from a filepath, populating missing items and
+        sections
+        Args:
+            config_file (str): the file to be parsed
+        """
+        # If the config file doesn't exist, prepopulate the config object
+        # with the defaults, in the right section.
+        #
+        # Otherwise, we have to put the defaults in the DEFAULT section,
+        # to ensure that they don't override anyone's settings which are
+        # in their config file in the default section (which is likely,
+        # because sydent used to be braindead).
+        use_defaults = not os.path.exists(config_file)
+        cfg = configparser.ConfigParser()
+        for sect, entries in CONFIG_DEFAULTS.items():
+            cfg.add_section(sect)
+            for k, v in entries.items():
+                cfg.set(configparser.DEFAULTSECT if use_defaults else sect, k, v)
+
+        cfg.read(config_file)
+
+        # Logging is configured in cfg, but these options must be parsed first
+        # so that we can log while parsing the rest
+        setup_logging(cfg)
+
+        for section in self.config_sections:
+            section.parse_config(cfg)
+
+            # Changes may need to be saved back to file (e.g. generated keys)
+            if hasattr(section, "update_cfg") and section.update_cfg:
+                fp = open(config_file, "w")
+                cfg.write(fp)
+                fp.close()
+
+    def parse_config_dict(self, config_dict: Dict):
+        """Parse the given config from a dictionary, populating missing items and sections
+
+        Args:
+            config_dict (dict): the configuration dictionary to be parsed
+        """
+        # Build a config dictionary from the defaults merged with the given dictionary
+        config = copy.deepcopy(CONFIG_DEFAULTS)
+        for section, section_dict in config_dict.items():
+            if section not in config:
+                config[section] = {}
+            for option in section_dict.keys():
+                config[section][option] = config_dict[section][option]
+
+        # Build a ConfigParser from the merged dictionary
+        cfg = configparser.ConfigParser()
+        for section, section_dict in config.items():
+            cfg.add_section(section)
+            for option, value in section_dict.items():
+                cfg.set(section, option, value)
+
+        # This is only ever called by tests so don't configure logging
+        # as tests do this themselves
+
+        for section in self.config_sections:
+            section.parse_config(cfg)
+
+
+def setup_logging(cfg: configparser.ConfigParser):
+    """
+    Setup logging using the options selected in the config
+
+    Args:
+        cfg (ConfigParser): the configuration
```

Comment:
Sydent doesn't use this format of docstring: rather it uses the sphinx format: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html.